File: ruby/creditcard_validation.rb
#!/usr/bin/ruby
#Credit card validation code in ruby
#Testcase is account number is 79927398713
def luhns(creditcardnumber)
	card_number = creditcardnumber.split("")
	card_number_evens = []
	card_number_odds = []
	card_number.each_with_index do |element, index|
		if index % 2 == 0
			element = element.to_i
			card_number_evens << element
		elsif index % 2 == 1
			element = element.to_i
			card_number_odds << element * 2
		end
	end
	#puts card_number_evens
	#puts "---------------------------"
	#puts card_number_odds
	#puts "---------------------------"
	card_number_odds.each_with_index do |element, index|
		if element > 9
			temparray  = element.to_s.split("")
			temparray.each_with_index do |number, nope|
				temparray[nope] = number.to_i
			end
			card_number_odds[index] = temparray.inject(:+)
			#puts card_number_odds	
			#puts "-------------------"
		end
	end
	final = (card_number_evens.concat(card_number_odds).inject(:+) * 9) % 10
	#puts final
	#puts "-------------------------"
	return final
	
	
end

puts "Enter the credit card number"
card_number = gets.chomp
digit = card_number[-1].to_i
#puts "digit is " , digit
if luhns(card_number[0..-2]) == digit
	puts "Valid"	
else
	puts "Invalid"
end

File: python/creditcard_validation.py
#!/usr/bin/python
#Test case is account number 79927398713
#Credit card validation algorithm availbale on Wikipedia

def luhns(card_number):
	card_number_ints = [int(x) for  x in card_number]
	even_places = card_number_ints[-1::-2]
	#print even_places[::-1]
	odd_places = card_number_ints[-2::-2]
	#print odd_places[::-1]
	double_evens = [x * 2 for x in even_places]
	#print double_evens[::-1]
	for i in range(len(double_evens)):
		if double_evens[i] > 9:
	#		print "Entered if"
			#print [int(x) for x in list(str(double_evens[i]))]
			double_evens[i] = sum([int(x) for x in list(str(double_evens[i]))])
	#print double_evens[::-1]
	final = (sum(odd_places + double_evens) * 9) % 10
	return final
def main():
	card_number = raw_input("Enter Account Number")
	if int(card_number[-1]) == luhns(card_number[:-1]):
		print "Valid card by mod10 algorithm"
	else:
		print "Nopenopenopenopenopenopenope"	 

if __name__ == "__main__":
	main()
